sarad/models/vae/utils.py

```diff
@@ -0,0 +1,247 @@
+from tensorflow.keras.callbacks import EarlyStopping
+from tensorflow.keras.layers import Dropout
+from tensorflow.keras.layers import Conv2D, MaxPooling2D, UpSampling2D, Dropout, BatchNormalization
+import matplotlib.pyplot as plt
+import pandas as pd
+from tensorflow.keras import Model, Input
+from tensorflow.keras.layers import Lambda
+from tensorflow.keras.layers import Flatten
+from tensorflow.keras.layers import Dense
+from tensorflow.keras.layers import Dropout, Conv2D, MaxPooling2D, UpSampling2D, BatchNormalization, Flatten, Dense, Input, Lambda, Reshape
+import tensorflow.keras.backend as K
+from tensorflow.keras.saving import register_keras_serializable
+from tensorflow.keras.layers import Conv2D, MaxPooling2D, UpSampling2D, Dense, Flatten, Dropout, Input, Reshape, Conv2DTranspose, Layer
+import tensorflow as tf
+from sklearn.preprocessing import LabelEncoder
+from sklearn.metrics.pairwise import cosine_distances
+import matplotlib.pyplot as plt
+import numpy as np
+
+class Sampling(Layer):
+    """Sampling layer using (mean, log_var)"""
+
+    def call(self, inputs):
+        mean, log_var = inputs
+        batch = tf.shape(mean)[0]
+        dim = tf.shape(mean)[1]
+        epsilon = tf.random.normal(shape=(batch, dim))
+        return mean + tf.exp(0.5 * log_var) * epsilon
+
+class Utils:
+    def __init__(self):
+        pass
+
+    def plot_history(self, history):
+        plt.plot(history.history['loss'], label='Train Loss')
+        plt.plot(history.history['val_loss'], label='Val Loss')
+        plt.title("Autoencoder Training Loss")
+        plt.xlabel("Epoch")
+        plt.ylabel("MSE Loss")
+        plt.legend()
+        plt.grid(True)
+        plt.tight_layout()
+        plt.show()
+
+    reconstructions = []
+
+    def evaluate_set(self, dataset=None, set_name="validation", valid_items=None, threshold=None, model=None, patches = None):
+        global reconstructions
+        # if dataset is None:
+        #     dataset = X_val
+        #     set_name = "validation"
+
+        if valid_items is None:
+            raise ValueError("⚠️ Must provide `valid_items` to match patch metadata.")
+
+        print(f"📊 Evaluating {set_name} set reconstructions...")
+        reconstructions = model.predict(dataset)
+
+        stats = []
+
+        for idx in range(len(dataset)):
+            original = dataset[idx]
+            recon = reconstructions[idx]
+            meta = patches[idx]
+            patch_id = meta.get("patch_id", f"{set_name}_{idx}")
+            true_label = int(meta.get("is_anomaly", False))
+
+            # Compute loss metrics
+            bce = tf.keras.losses.binary_crossentropy(
+                tf.convert_to_tensor(original.flatten()),
+                tf.convert_to_tensor(recon.flatten())
+            ).numpy().mean()
+
+            mse = np.mean((original - recon) ** 2)
+            mae = np.mean(np.abs(original - recon))
+            diff = np.abs(original - recon)
```

Comment:
```suggestion
            diff = np.abs(original - recon)
            mae = np.mean(diff)
```
Because why wouldn't you?